src/main/java/com/onaio/steps/helper/UploadFileTask.java

```diff
@@ -67,36 +68,53 @@ public void upload(File file) {
-                    @Override
-                    public void onResponse(Call<ResponseBody> call, Response<ResponseBody> response) {
-                        if (response.isSuccessful() && response.code() == 201) {
-                            new CustomNotification().notify(activity, R.string.export_complete, R.string.export_complete_message);
-                            onExportListener.onFileUploaded(true);
-                        } else {
-                            new CustomNotification().notify(activity, R.string.error_title, R.string.export_failed);
-                            onExportListener.onFileUploaded(false);
-                        }
-                    }
-
-                    @Override
-                    public void onFailure(Call<ResponseBody> call, Throwable t) {
-                        new CustomNotification().notify(activity, R.string.error_title, R.string.export_failed);
-                        onExportListener.onFileUploaded(false);
-                    }
-                });
+                if (!files.isEmpty()) {
+                    upload(files, endPoint, surveyIdBody, userIdBody, userPasswordBody);
+                }
             }
         } else {
             new CustomNotification().notify(activity, R.string.error_title, R.string.export_failed);
             onExportListener.onFileUploaded(false);
         }
     }
+
+    public void upload(Queue<File> files, String endPoint, RequestBody surveyIdBody, RequestBody userIdBody, RequestBody userPasswordBody) {
+
+        File file = files.remove();
+        String fileType = "text/csv";
+        RequestBody requestFile = RequestBody.create(MediaType.parse(fileType), file);
+        MultipartBody.Part fileBody = MultipartBody.Part.createFormData("file", file.getName(), requestFile);
+
+        retrofit.create(HouseholdService.class).uploadData(endPoint, fileBody, surveyIdBody, userIdBody, userPasswordBody).enqueue(new Callback<ResponseBody>() {
+            @Override
+            public void onResponse(Call<ResponseBody> call, Response<ResponseBody> response) {
+                if (response.isSuccessful() && response.code() == 201) {
+                    if (files.isEmpty()) {
+                        new CustomNotification().notify(activity, R.string.export_complete, R.string.export_complete_message);
+                        onExportListener.onFileUploaded(true);
+                    } else {
+                        upload(files, endPoint, surveyIdBody, userIdBody, userPasswordBody);
+                    }
+                } else {
+                    if (files.isEmpty()) {
+                        new CustomNotification().notify(activity, R.string.error_title, R.string.export_failed);
+                        onExportListener.onFileUploaded(false);
+                    }
+                    else {
+                        upload(files, endPoint, surveyIdBody, userIdBody, userPasswordBody);
```

Comment:
This does not handle failed file uploads correctly since they will end up being marked as uploaded

Reply:
Basically, we can end up having missing data on the server and cannot clearly tell what was not sent on inspecting the steps app database

Reply:
https://github.com/onaio/steps-app/pull/158#discussion_r840647213